hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/DatanodeInfo.java

```diff
@@ -286,11 +286,20 @@ public void setNodeStatus(NodeStatus newNodeStatus) {
    * Set the current command counts for this datanode, as reported in the last
    * heartbeat.
    * @param cmds Proto message containing a list of command count pairs.
+   * @param commandsToBeSent Summary of commands which will be sent to the DN
+   *                         as the heartbeat is processed and should be added
+   *                         to the command count.
    */
-  public void setCommandCounts(CommandQueueReportProto cmds) {
+  public void setCommandCounts(CommandQueueReportProto cmds,
+      Map<SCMCommandProto.Type, Integer> commandsToBeSent) {
     try {
       int count = cmds.getCommandCount();
+      Map<SCMCommandProto.Type, Integer> mutableCmds
+          = new HashMap<>(commandsToBeSent);
       lock.writeLock().lock();
+      // Purge the existing counts, as each report should completely replace
+      // the existing counts.
+      commandCounts.clear();
```

Comment:
Just a quick question, why was this not done before and done now? I understand we would like to keep the current count and replace the stale counters. I thought it was already done by `commandCounts.put(command, cmdCount)` and also now with `commandCounts.putAll(mutableCmds)`.

Reply:
I'm not sure I follow your question.

There are two counts. The ones reported from the DN on its heartbeart - these are queued on the DN.

Then there are commands queued in SCM ready to be sent to the DN as part of processing the heartbeat.

After the heartbeat, the real queue on the DN is the sum of the two (queued on DN + queued on SCM).

Before this change, we only stored the "queued on the DN count". This count is not yet used anywhere, but I missed adding the commands the heartbeat will add to the queue, so the old count is not correct.

The line `commandCounts.put(command, cmdCount)` should do all the updates as per the comment in the code. The line `commandCounts.putAll(mutableCmds)` is a "catch all" incase the DN is not reporting counts for a command that is queued on SCM - it shouldn't happen in practice. I have also removed the entry from mutableCommands after adding it

Reply:
Yea not sure why I didn't have that before. I guess we could omit it, but I should not do any harm either.